[clr-interp] Interpreter for ARM32 SOFTFP

clamp03 opened the pull request:
Enable interpreter for arm32 softfp.
- Implement Assemblies for args and return values
- Fix some minor bugs
- Tested with simple test cases

clamp03 committed:
[CLRINTRP] Port to ARM

clamp03 committed:
[CLRINTRP] Make dummy asssembly functions

clamp03 committed:
Disable Float

clamp03 committed:
[INTRP] Print HelloWorld Done

However, it crashes right after HellWorld print.

clamp03 committed:
[INTRP] HelloWorld Done

clamp03 committed:
[INTRP] Update X to R for arm register

clamp03 committed:
[INTRP] Call methods with 2 args and return value

clamp03 committed:
[INTRP] Implement Args Load and Store Assemblies

clamp03 committed:
[INTRP] Add Load_Ref and Store_Ref

clamp03 committed:
[INTRP] Return I8

clamp03 committed:
[INTRP] Update I8 to I4 for ARM32

clamp03 committed:
[INTRP] Call to Method with LONG args

clamp03 committed:
[INTRP] Initial Support for Float and Double

clamp03 committed:
[INTRP] Update for ARM SOFTFP

clamp03 committed:
[INTRP] Update all assemblies for arm softp

Need to test the implementations

clamp03 committed:
[INTRP] Can Pass Arguments using Stack

clamp03 committed:
[INTRP] Remove Redundant Codes

clamp03 committed:
[INTRP] Support Struct

clamp03 committed:
[INTRP] Support RetBuff

clamp03 committed:
[DO NOT MERGE] For Test

clamp03 committed:
[INTRP] Fix release build errors

clamp03 committed:
Revert "[DO NOT MERGE] For Test"

This reverts commit b3ba6ec68f377f36b7c8dcfda40acf72a1647a93.

clamp03 committed:
Workarounds for errors

clamp03 committed:
Revert INTERP_STACK_SLOT_SIZE

clamp03 commented on src/coreclr/interpreter/inc/interpretershared.h:
```diff
@@ -14,7 +14,11 @@
+#ifdef TARGET_64BIT
 #define INTERP_STACK_SLOT_SIZE 8    // Alignment of each var offset on the interpreter stack
+#else // !TARGET_64BIT
+#define INTERP_STACK_SLOT_SIZE 4    // Alignment of each var offset on the interpreter stack
```
@kg I found a problem when I change it to 8 byte-stack slot. (Actually, I forgot implementation details during my long holidays. :smiling_face_with_tear:)

If I change it to 8 byte-stack slot, it seems passing args between compiled methods and interpreter is hard. When it passes two 4-bytes args or one 8-bytes arg, it uses 2 registers in ARM32. So if it is one 8-bytes arg, values in two registers are needed to be loaded from or stored to `one stack slot`. However in case of two 4-bytes args, values in two registers are loaded from or stored to `two stack slots`. In current implementation, argument passing is handled by Load_* and Store_* routines in assembly code without any type check. However, if stack slot is 8 bytes, I think it needs to do type check for all args (or make routines for all cases.). 
How do you solve this in mono interpreter? Could you share any idea about this?
Thank you.

clamp03 committed:
Revert "Revert INTERP_STACK_SLOT_SIZE"

This reverts commit 4d48c033cc39aea0d93a14a3e57ce5a87cb6de7d.

clamp03 committed:
Update StackVal for 32 bit target

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Fix align for 8 bytes and value type

clamp03 committed:
8 bytes stack size for WASM

clamp03 commented on src/coreclr/interpreter/inc/interpretershared.h:
```diff
@@ -14,11 +14,11 @@
 #endif // _MSC_VER
 
-#ifdef TARGET_64BIT
+#if defined(TARGET_64BIT) || defined(TARGET_WASM)
```
Sorry, I am not sure what you mean. From my understanding, you want to use the same 8-byte interpreter stack alignment plan for all platforms like @kg mentioned earlier. Is it correct?

I think I can make arm32 on 8-byte interpreter stack by adding more routines about 8 bytes value and 4 bytes value.
Thank you.

clamp03 committed:
Change interpreter stack size to 8 bytes

- Handle copying args / ret value between interpreter stack and native stack
- No Range Expansion for value (>= 8 bytes)
- Terminate current routines and add a routine for the value

clamp03 committed:
Update ProcessArgument

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Fix InterpreterStub

kg commented:
> @kg I tried Interpreter.cs. TestPreciseInitCctors test in Interpreter.cs fails with error message: `preciseInitCctorsRun should be 1, but is 0` when I give any `DOTNET_Interpreter` option like `DOTNET_Interpreter=none`. It fails on x64 as well as arm32. Could you share how you are testing with it?

The correct way to run Interpreter.cs is either by running InterpreterTester (it sets the environment variables) or by setting all the appropriate environment variables correctly.

IIRC you need to set a few vars - disable readytorun, disable tiered compilation, and force on the interpreter for everything inside the Interpreter.dll module. Are you using InterpreterTester? If so, it sounds like maybe your changes have caused a regression somehow. Does it work for you on latest main?

clamp03 committed:
Fix errors in arm assembly code

clamp03 committed:
Fix stack value

Sync stack value between interpreter frame and native stack for
unwinding

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Use TARGET_32BIT instead of TARGET_64BIT

clamp03 committed:
InterpreterTest

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Update OFFSETOF__Thread__m_pInterpThreadContext

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 commented:
@janvorli @kg @jkotas Could you review this PR?

janvorli commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -378,7 +378,11 @@ typedef void (*HELPER_FTN_V_PP)(void*, void*);
 {
     // FIXME VirtualAlloc/mmap with INTERP_STACK_ALIGNMENT alignment
+#ifdef TARGET_UNIX
+    pStackStart = pStackPointer = (int8_t*)aligned_alloc(INTERP_STACK_ALIGNMENT, INTERP_STACK_SIZE);
```
Coreclr itself doesn't call Unix specific functions. This should be hidden in the minipal. Historically, PAL was used for this stuff and we are trying to reduce it. The minipal is the new place for abstracting platform dependencies. 

Can you please add a new method for aligned allocation to the VMToOSInterface in https://github.com/dotnet/runtime/blob/3345865f57f59f4a54e1749c06f5a2782d961a16/src/coreclr/minipal/minipal.h and implement it in files in the related Unix and Windows subfolders? I would recommend adding a new source file named "memory.cpp" there. In the future, I would add support for memory mapping functionality there and remove that from the PAL.

Also, this would not work on macOS, it should use `posix_memalign` there instead.

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Update

jkotas commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -2522,7 +2526,14 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
                     {
+#ifdef TARGET_ARM
+                        TADDR targetSP = pInterpreterFrame->GetInterpExecMethodSP();
+                        pInterpreterFrame->SetInterpExecMethodSP(targetSP - 8); // Pass two arguments via native stack for InvokeUnmanagedMethodWithTransition
```
`InvokeUnmanagedMethodWithTransition` is marked with NOINLINING and packs all arguments into a struct for other reasons, so it would not be that bad to pack the arguments into a struct here.

clamp03 replied:
Thank you. I updated argument passing. I am not sure I implemented as what you intended. Could you review the change? Thank you.

clamp03 committed:
Revert STACK_ALIGN_SIZE

clamp03 committed:
Pass all arguments by a struct

For InvokeUnmanagedMethodWithTransition, pass all arguments by a struct.

clamp03 committed:
Move target specific function to minipal

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 commented on src/coreclr/vm/arm/cgencpu.h:
```diff
@@ -62,7 +62,7 @@ EXTERN_C void checkStack(void);
 #define COMMETHOD_PREPAD                        12   // # extra bytes to allocate in addition to sizeof(ComCallMethodDesc)
 
-#define STACK_ALIGN_SIZE                        4
+#define STACK_ALIGN_SIZE                        8
```
Thank you. I reverted.

jkotas commented on src/coreclr/minipal/Windows/memory.cpp:
```diff
@@ -0,0 +1,20 @@
+
+void VMToOSInterface::AlignedFree(void *memblock)
+{
+    free(memblock);
```
```suggestion
    _aligned_free(memblock);
```

jkotas commented on src/coreclr/minipal/Unix/memory.cpp:
```diff
@@ -0,0 +1,24 @@
+#include <inttypes.h>
+#include <stdlib.h>
+#include "minipal.h"
+#include "minipal/cpufeatures.h"
```
```suggestion
```

clamp03 committed:
Fix errors

jkotas commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -88,6 +120,7 @@ void InvokeUnmanagedMethodWithTransition(MethodDesc *targetMethod, int8_t *stack
     param = { targetMethod, pArgs, pRet, callTarget };
+#endif // TARGET_ARM
 
     PAL_TRY(Param *, pParam, &param)
```
```suggestion
    PAL_TRY(UnmanagedMethodWithTransitionParam *, pParam, pParam)
```
And then use the same structure as the parameter for PAL_TRY macro

clamp03 committed:
Fix: Apply review feedback

clamp03 committed:
Remove a header inclusion

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Updated based on reviews

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 commented on src/coreclr/clrfeatures.cmake:
```diff
@@ -46,11 +46,11 @@ if(NOT DEFINED FEATURE_INTERPRETER)
+    else(CLR_CMAKE_TARGET_ARCH_AMD64 OR CLR_CMAKE_TARGET_ARCH_ARM64 OR CLR_CMAKE_TARGET_ARCH_RISCV64 OR CLR_CMAKE_TARGET_ARCH_ARM)
       set(FEATURE_INTERPRETER 0)
-    endif(CLR_CMAKE_TARGET_ARCH_AMD64 OR CLR_CMAKE_TARGET_ARCH_ARM64 OR CLR_CMAKE_TARGET_ARCH_RISCV64)
+    endif(CLR_CMAKE_TARGET_ARCH_AMD64 OR CLR_CMAKE_TARGET_ARCH_ARM64 OR CLR_CMAKE_TARGET_ARCH_RISCV64 OR CLR_CMAKE_TARGET_ARCH_ARM)
```
Thank you.

janvorli commented on src/coreclr/vm/callstubgenerator.cpp:
```diff
@@ -2416,7 +2599,10 @@ void CallStubGenerator::TerminateCurrentRoutineIfNotOfNewType(RoutineType type,
     }
-    else if ((m_currentRoutineType == RoutineType::FPReg) && (type != RoutineType::FPReg))
+#ifndef TARGET_ARM
+    else
```
I don't understand why this ifdef is needed.

janvorli replied:
Also, the case when a struct is split between registers and stack also exists on x64 Linux. The ArgIterator has the GetArgLocDescForStructInRegs() for this purpose. The code in the ComputeCallStub uses that to pretend the parts of the struct are "separate arguments" from the further processing point of view. The stuff around it is quite complex due to the complicated classification mechanism defined by the Unix amd64 ABI, but it seems we could handle the case of args split between regs and stack at that place too and I think we would not have to handle that here then.

janvorli commented on src/coreclr/vm/callstubgenerator.cpp:
```diff
@@ -2436,10 +2622,18 @@ void CallStubGenerator::TerminateCurrentRoutineIfNotOfNewType(RoutineType type,
 #endif // TARGET_ARM64
-    else if ((m_currentRoutineType == RoutineType::Stack) && (type != RoutineType::Stack))
+#ifndef TARGET_ARM
+    else
```
The same here

janvorli commented on src/coreclr/vm/callstubgenerator.cpp:
```diff
@@ -2645,21 +2848,32 @@ void CallStubGenerator::ProcessArgument(ArgIterator *pArgIt, ArgLocDesc& argLocD
         else if (argLocDesc.m_idxGenReg == m_r2 + 1 && (!pArgIt || !pArgIt->IsArgPassedByRef()))
+#else
+        else if (argLocDesc.m_idxGenReg == m_r2 + 1)
+#endif // ENREGISTERED_PARAMTYPE_MAXSIZE
```
A nit - can you please keep only the extra part of the condition under the ifdef here? Like this:
```suggestion
        else if (argLocDesc.m_idxGenReg == m_r2 + 1 
#ifdef ENREGISTERED_PARAMTYPE_MAXSIZE
        && (!pArgIt || !pArgIt->IsArgPassedByRef())
#endif
        )
```

janvorli commented on src/coreclr/vm/callstubgenerator.cpp:
```diff
@@ -2730,13 +2944,26 @@ void CallStubGenerator::ProcessArgument(ArgIterator *pArgIt, ArgLocDesc& argLocD
         }
-        else if ((argLocDesc.m_byteStackIndex == m_s2 + 1) && (argLocDesc.m_byteStackSize >= 8) && (!pArgIt || !pArgIt->IsArgPassedByRef()))
+#ifdef ENREGISTERED_PARAMTYPE_MAXSIZE
+        else if ((argLocDesc.m_byteStackIndex == m_s2 + 1) && (argLocDesc.m_byteStackSize >= TARGET_POINTER_SIZE) && (!pArgIt || !pArgIt->IsArgPassedByRef()))
```
A nit - can you please keep only the extra part of the condition under the ifdef here?

janvorli commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -68,28 +68,29 @@ std::invoke_result_t<Function> CallWithSEHWrapper(Function function)
 // Use the NOINLINE to ensure that the InlinedCallFrame in this method is a lower stack address than any InterpMethodContextFrame values.
 NOINLINE
-void InvokeUnmanagedMethodWithTransition(MethodDesc *targetMethod, int8_t *stack, InterpMethodContextFrame *pFrame, int8_t *pArgs, int8_t *pRet, PCODE callTarget)
+void InvokeUnmanagedMethodWithTransition(UnmanagedMethodWithTransitionParam *pParam)
```
Can you please add a brief comment explaining why we are passing the args via this struct instead of directly so that people can understand the reason in the future?

clamp03 replied:
I added a comment. If it isn't good enough, I will update again. Thank you.

clamp03 committed:
Updated based on reviews

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Update special handling for arm32

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Fix errors after upstream merge

- Because of async support, some methods need arguments more than 4.
  I changed arguments passing.
- Fixed some build errors.

clamp03 committed:
Fix build error

janvorli commented on src/coreclr/vm/callstubgenerator.cpp:
```diff
@@ -2661,21 +2863,32 @@ void CallStubGenerator::ProcessArgument(ArgIterator *pArgIt, ArgLocDesc& argLocD
+#ifdef TARGET_ARM
+        if (argLocDesc.m_cGenReg == 2)
+        {
+            pRoutines[m_routineIndex++] = GetRegRoutine_4B(argLocDesc.m_idxGenReg, argLocDesc.m_idxGenReg + argLocDesc.m_cGenReg - 1);
```
Have you considered / tried changing the interpreter stack so that on arm32, the slot size would be 32 bit? It seems it would then eliminate the need for these special _4B variants. However, I am not sure if there are any blockers / issues with changing the stack slot size. I've asked in a private discussion why Mono doesn't use 32 bit interpreter stack on arm32 and the answer was that there was no reason for trying to do that there.

clamp03 replied:
> Have you considered / tried changing the interpreter stack so that on arm32, the slot size would be 32 bit? It seems it would then eliminate the need for these special _4B variants. However, I am not sure if there are any blockers / issues with changing the stack slot size. I've asked in a private discussion why Mono doesn't use 32 bit interpreter stack on arm32 and the answer was that there was no reason for trying to do that there.

At first, I implemented with 32 bit slot size and the tests passed because I thought 32 bit slot size is better in 32 bit arch. After https://github.com/dotnet/runtime/pull/120688#discussion_r2446575365, I changed it to 64 bit. I am sure which is better for .NET runtime interpreter. Could you please let me know your idea?
Thank you.

janvorli replied:
Let's keep it as is, we may revisit it later if it turns out that 32 bit stack would be better for WASM which is also 32 bit.

clamp03 committed:
Fix a mistake

clamp03 commented on src/coreclr/vm/callstubgenerator.cpp:
```diff
@@ -2636,6 +2831,13 @@ void CallStubGenerator::ProcessArgument(ArgIterator *pArgIt, ArgLocDesc& argLocD
+#ifdef TARGET_ARM
+    if (argLocDesc.m_cGenReg == 2 || argLocDesc.m_byteStackSize >= 8)
+    {
+        /* do nothing */
```
> But single routine handles only a stack or registers. For the case you have described, there would need to be two routines. In the current method, this added ifdef seems to have no effect. If the m_currentRoutineType was RoutineType::GPReg, then it would not enter the if below with or without the else. And if the m_currentRoutineType was RoutineType::FPReg, then again, the else would have no effect.

I am so sorry. When I have implemented it in the first time, removing `else` statements was valid . After the `TerminateCurrentRoutineIfNotOfNewType` function is introduced, I didn't read codes carefully.I will check `GetArgLocDescForStructInRegs` which you mentioned in another comment.

> Can you please help me understand why arguments with 2 registers or 8 or more bytes passed on stack don't get any routine classification here?
The goal of this part of the code together with the TerminateCurrentRoutineIfNotOfNewType call is to identify when we cannot extend the range of registers / stack that we copy using a single routine. So with this change, any time we get an argument that is passed in two regs or an argument that is passed on stack with 8 or more bytes, we would not extend the previous range of registers / stack and I am not sure why.

I intended to block to extend of the previous ranges for long bytes and to enable to extend 4-size single register / stack. Because for 4-byte arg and 8-byte arg are handled differently. For 4 byte, it load a single interp stack to a register or stack. For 8 bytes, it load a single interp stack to two registers or stacks. If 4-byte and 8-byte args are mixed in a range, I think it is hard to identify how to handle.

clamp03 replied:
Removed the def in `TerminateCurrentRoutineIfNotOfNewType`. And for arg which has both reg and stack, execute `ProcessArgument` for each. Could you review again?

clamp03 committed:
Update ComputeCallStub

- To support structure arg which has reg and stack, split it for each

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Fix offset

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

janvorli reviewed (commented):
LGTM, I just have a suggestion on updating a comment. 

Before merging it, could you please update the call stub generator and the asm helpers to fix the problem I've fixed yesterday for other architectures in #123402?

clamp03 committed:
Merge remote-tracking branch 'origin/main' into intrp_arm

clamp03 committed:
Replace code comment by suggested change

clamp03 committed:
Add small int return value conversion to arm

janvorli reviewed (approved):
LGTM, thank you!
